altapay.php

```diff
@@ -0,0 +1,2389 @@
+        }
+        $voucherProperties['freeShippingStatus'] = $freeShipping;
+
+        return $voucherProperties;
+    }
+
+    /**
+     * Returns the array with products in cart rule group along with the reduction percentage
+     *
+     * @param int $couponID
+     * @param int $reductionPercent
+     *
+     * @return array
+     */
+    private function getCartRuleGroupProducts($couponID, $reductionPercent)
+    {
+        $cartRuleGroupProducts = [];
+        $cartRuleGroups        = Db::getInstance()->executeS('SELECT * FROM ' . _DB_PREFIX_ . 'cart_rule_product_rule_group WHERE id_cart_rule = ' . $couponID);
+        foreach ($cartRuleGroups as $cartRuleGroup) {
+            $cartRuleGroupProducts['reductionPercent'] = $reductionPercent;
+            $cartRuleGroupProducts['products'] = $this->getCartRuleGroupProductIDs($cartRuleGroup['id_product_rule_group']);
+        }
+
+        return $cartRuleGroupProducts;
+    }
+
+    /**
+     * Return IDs of the products in cart rule group
+     *
+     * @param int $cartRuleGroupID
+     *
+     * @return array
+     */
+    private function getCartRuleGroupProductIDs($cartRuleGroupID)
+    {
+        $productIDs     = [];
+        $cartRuleGroups = Db::getInstance()->executeS('SELECT * FROM ' . _DB_PREFIX_ . 'cart_rule_product_rule_value WHERE id_product_rule = ' . $cartRuleGroupID);
+        foreach ($cartRuleGroups as $cartRuleGroup) {
+            $productIDs[] = $cartRuleGroup['id_item'];
+        }
+
+        return $productIDs;
+    }
+
+    /**
+     * Returns array of applied voucher details from cart
+     *
+     * @return array
+     */
+    private function getVoucherDetails()
+    {
+        $voucherDetails   = [];
+        $appliedCartRules = $this->context->cart->getCartRules();
+        foreach ($appliedCartRules as $cartRule) {
+            $reductionPercent = $cartRule['reduction_percent'];
+            if (!empty($cartRule['reduction_product'])) {
+                $voucherDetails[$cartRule['id_cart_rule']] = $this->getCartRuleGroupProducts($cartRule['id_cart_rule'], $reductionPercent);
+            } else {
+                $voucherDetails[$cartRule['id_cart_rule']] = [
+                    'reductionPercent' => $reductionPercent,
+                    'products'         => 'all'
+                ];
+            }
+        }
+
+        return $voucherDetails;
+    }
+
+    /**
+     * Returns array of cart rule discounts applied on each product from created order
+     *
+     * @param array $order
+     *
+     * @return array
+     */
+    private function getCartRuleDiscounts($order)
+{
+    $cartRuleDiscounts = [];
+    $discountPercent   = reset(Db::getInstance()->executeS('SELECT * FROM ' . _DB_PREFIX_ . 'altapay_cartInfo WHERE id_cart = ' . $order->id_cart));
+
+      if (!empty($discountPercent) && isset($discountPercent['productDetails'])) {
+        $cartRuleDiscounts = json_decode($discountPercent['productDetails']);
```

Comment:
We need to handle cases where json_decode fails. Previously the outer functions would handle this was handled by the caller by doing nothing if the result was falsy. We can maintain this behavior by returning an empty array, or we could be more explicit and throw an exception.
```suggestion
        $cartRuleDiscounts = json_decode($discountPercent['productDetails']) ? [];
```